group.go

```diff
@@ -23,7 +34,48 @@ func (g *Group) NewGroup(path string) *Group {
 	if path[len(path)-1] == '/' {
 		path = path[:len(path)-1]
 	}
-	return &Group{path, g.mux}
+	return &Group{
+		path:  path,
+		mux:   g.mux,
+		stack: g.stack,
+	}
+}
+
+// Use appends a middleware handler to the Group middleware stack.
+func (g *Group) Use(fn MiddlewareFunc) {
+	g.stack = append(g.stack, fn)
+}
+
+// UseHandlerFunc is like Use but accepts http.HandlerFunc.
+func (g *Group) UseHandlerFunc(fn http.HandlerFunc) {
+	g.stack = append(g.stack, func(next HandlerFunc) HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request, params map[string]string) {
+			fn(w, r)
+			next(w, r, params)
+		}
+	})
+}
+
+type handlerWithParams struct {
+	handler HandlerFunc
+	params  map[string]string
+}
+
+func (h handlerWithParams) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.handler(w, r, h.params)
+}
+
+// UseHandler is like Use but accepts http.Handler middleware.
+func (g *Group) UseHandler(middleware func(http.Handler) http.Handler) {
+	g.stack = append(g.stack, func(next HandlerFunc) HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request, params map[string]string) {
+			nextHandler := handlerWithParams{
+				handler: next,
+				params:  params,
+			}
+			middleware(nextHandler).ServeHTTP(w, r)
```

Comment:
Hard to tell how good this will work in practice if middleware does some heavy initialization on `middleware(nextHandler)` call.

Reply:
IMHO, middleware initialization should be lightweight, ie:

```golang
package middleware

type Middleware struct {...}

func New(...) *Middleware { /* initialize here */ }

type handler {
    m    *Middleware
    next http.Handler
}

// implement handler as http.Handler

func (m *Middleware) Handler (next func(http.Handler) http.Handler {
    return handler{m: m, next: next}
}
```

Reply:
One big reason for it to be lightweight is, that middleware will be used with many routes (different `http.Handler`s), and therefore it doesn't make sense to do heavy initialization for each route separately.

Reply:
func(http.Handler) http.Handler)

Reply:
Though,... this is internal implementation. The primary focus should be on public/exported API, which should stay same, once it's released.

Reply:
Not sure I follow. I mean that some existing 3rd party middlewares may do some initialization on `middleware(nextHandler)` call since most routers call `middleware(nextHandler)` once - not for every route call. We can't do that because we need to capture `params`.

Reply:
> ... call since most routers call `middleware(nextHandler)` once - not for every route call...

Not once per handled request, but once per route handler registration, as each registered route has got different handler.

With tens of registered routes, it doesn't make sense to have heavy weight initialization there, as it would mean for middleware to have heavy-weight handler wrapper,... therefore probably lots of data duplication (unnecessary memory consumption).